sockeye/rnn_attention.py

```diff
@@ -343,33 +364,65 @@ def attend(att_input: AttentionInput, att_state: AttentionState) -> AttentionSta
             attention_scores = mx.sym.batch_dot(lhs=source_hidden, rhs=expanded_decoder_state,
                                                 name="%sbatch_dot" % self.prefix)
 
-            context, attention_probs = get_context_and_attention_probs(source, source_length, attention_scores)
+            context, attention_probs = get_context_and_attention_probs(source, source_length, attention_scores, self.dtype)
             return AttentionState(context=context,
                                   probs=attention_probs,
                                   dynamic_source=att_state.dynamic_source)
 
         return attend
 
 
+@Attention.register(C.ATT_DOT_SCALED)
+class DotScaledAttention(DotAttention):
+    """
+    DotAttention with optional scale before dot product [Vaswani et al, 2017].
+
+    :param input_previous_word: Feed the previous target embedding into the attention mechanism.
+    :param source_num_hidden: Number of hidden units in source.
+    :param query_num_hidden: Number of hidden units in query.
+    :param num_hidden: Number of hidden units.
+    :param scale: Optionally scale query before dot product [Vaswani et al, 2017].
+    :param dtype: data type.
+    """
+
+    def __init__(self,
+                 input_previous_word: bool,
+                 source_num_hidden: int,
+                 query_num_hidden: int,
+                 num_hidden: int,
+                 scale: Optional[float] = None,
+                 dtype: str = C.DTYPE_FP32) -> None:
+        super().__init__(input_previous_word,
+                         source_num_hidden,
+                         query_num_hidden,
+                         num_hidden,
+                         dtype=dtype)
+        self.scale = scale if scale else num_hidden ** -0.5
+
+
+@Attention.register(C.ATT_MH_DOT)
 class MultiHeadDotAttention(Attention):
     """
     Dot product attention with multiple heads as proposed in Vaswani et al, Attention is all you need.
     Can be used with a RecurrentDecoder.
 
     :param input_previous_word: Feed the previous target embedding into the attention mechanism.
-    :param num_hidden: Number of hidden units.
-    :param heads: Number of attention heads / independently computed attention scores.
+    :param source_num_hidden: Number of hidden units.
+    :param num_heads: Number of attention heads / independently computed attention scores.
+    :param dtype: data type.
     """
 
     def __init__(self,
                  input_previous_word: bool,
-                 num_hidden: int,
-                 heads: int) -> None:
-        super().__init__(input_previous_word)
-        utils.check_condition(num_hidden % heads == 0,
-                              "Number of heads (%d) must divide attention depth (%d)" % (heads, num_hidden))
-        self.num_hidden = num_hidden
-        self.heads = heads
+                 source_num_hidden: int,
```

Comment:
renaming these arguments means that models trained with the previous sockeye version will no longer be compatible. We should either leave the names the same, or bump (which might anyway be necessary if we decide to remove the scaled dot attention)

Reply:
I renamed constructor args, to match `AttentionConfig` state variables. Internal class fields as well as instantiation of `AttentionConfig` are untouched. So the change is backwards-compatible.

Reply:
oh ok. I mean there were anyway some backwards incompatible changes in the pipeline, so that we bump (see below), so either bumping shouldn't be a big issue.